core/src/main/scala/kafka/coordinator/group/CoordinatorPartitionWriter.scala

```diff
@@ -0,0 +1,171 @@
+  compressionType: CompressionType,
+  time: Time
+) extends PartitionWriter[T] {
+  // We use an action queue which directly executes actions. This is possible
+  // here because we don't hold any conflicting locks.
+  private val directActionQueue = new ActionQueue {
+    override def add(action: () => Unit): Unit = {
+      action()
+    }
+
+    override def tryCompleteActions(): Unit = {}
+  }
+
+  /**
+   * Register a PartitionWriter.Listener.
+   *
+   * @param tp       The partition to register the listener to.
+   * @param listener The listener.
+   */
+  override def registerListener(
+    tp: TopicPartition,
+    listener: PartitionWriter.Listener
+  ): Unit = {
+    replicaManager.maybeAddListener(tp, new ListenerAdapter(listener))
+  }
+
+  /**
+   * Deregister a PartitionWriter.Listener.
+   *
+   * @param tp       The partition to deregister the listener from.
+   * @param listener The listener.
+   */
+  override def deregisterListener(
+    tp: TopicPartition,
+    listener: PartitionWriter.Listener
+  ): Unit = {
+    replicaManager.removeListener(tp, new ListenerAdapter(listener))
+  }
+
+  /**
+   * Write records to the partitions. Records are written in one batch so
+   * atomicity is guaranteed.
+   *
+   * @param tp      The partition to write records to.
+   * @param records The list of records. The records are written in a single batch.
+   * @return The log end offset right after the written records.
+   * @throws KafkaException Any KafkaException caught during the write operation.
+   */
+  override def append(
+    tp: TopicPartition,
+    records: util.List[T]
+  ): Long = {
+    if (records.isEmpty) throw new IllegalStateException("records must be non-empty.")
+
+    replicaManager.getLogConfig(tp) match {
+      case Some(logConfig) =>
+        val magic = logConfig.recordVersion.value
+        val maxBatchSize = logConfig.maxMessageSize
+        val currentTimeMs = time.milliseconds()
+
+        val recordsBuilder = MemoryRecords.builder(
+          ByteBuffer.allocate(math.min(16384, maxBatchSize)),
+          magic,
+          compressionType,
+          TimestampType.CREATE_TIME,
+          0L,
+          maxBatchSize
+        )
+
+        records.forEach { record =>
+          val keyBytes = serializer.serializeKey(record)
+          val valueBytes = serializer.serializeValue(record)
+
+          if (recordsBuilder.hasRoomFor(currentTimeMs, keyBytes, valueBytes, EMPTY_HEADERS)) recordsBuilder.append(
+            currentTimeMs,
+            keyBytes,
+            valueBytes,
+            EMPTY_HEADERS
+          ) else throw new RecordTooLargeException(s"Message batch size is ${recordsBuilder.estimatedSizeInBytes()} bytes " +
+            s"in append to partition $tp which exceeds the maximum configured size of $maxBatchSize.")
+        }
+
+        var appendResults: Map[TopicPartition, PartitionResponse] = Map.empty
+        replicaManager.appendRecords(
+          timeout = 0L,
+          requiredAcks = 1,
```

Comment:
acks=1 is to not avoid going to the purgatory. as we don’t go the purgatory, the timeout does not matter.

Reply:
right, which means it's kind of a workaround and is confusing because the acks setting is unrelated to how we're using appendRecords as well as the timeout.

i'm wondering if we can set the default values in appendRecords for acks and timeout

Reply:
it is not a workaround. this is exactly what acks=1 is for, even for regular produce request. adding default values would be risky here. it is much better to let the caller be explicit about what he needs/wants.

Reply:
gotcha